substrate/utils/frame/benchmarking-cli/src/storage/read.rs

```diff
@@ -57,85 +69,185 @@ impl StorageCmd {
-						.build();
+				let is_batch_full = read_in_batch >= self.params.batch_size ||
+					(last_child_key == key && last_child_info == info);
+
+				// Read child keys on block validation
+				if is_batch_full && self.params.on_block_validation {
+					let root = backend.root();
+					let storage_proof = recorder
+						.clone()
+						.map(|r| r.drain_storage_proof())
+						.expect("Storage proof must exist for block validation");
+					let elapsed = measure_on_block_validation::<B>(
+						*root,
+						storage_proof,
+						on_validation_batch.clone(),
+						self.params.on_block_validation_rounds,
+					);
+					record.append(on_validation_size / on_validation_batch.len(), elapsed)?;
+
+					on_validation_batch = vec![];
+					on_validation_size = 0;
+				}
+
+				// Reload recorder
+				if is_batch_full {
+					(backend, recorder) = self.create_backend::<B, C>(&state);
 					read_in_batch = 0;
 				}
 			}
 		}
+
 		Ok(record)
 	}
+
+	fn create_backend<'a, B, C>(
+		&self,
+		state: &'a C::StateBackend,
+	) -> (
+		sp_state_machine::TrieBackend<
+			&'a <C::StateBackend as AsTrieBackend<HashingFor<B>>>::TrieBackendStorage,
+			HashingFor<B>,
+			&'a sp_trie::cache::LocalTrieCache<HashingFor<B>>,
+		>,
+		Option<sp_trie::recorder::Recorder<HashingFor<B>>>,
+	)
+	where
+		C: CallApiAt<B>,
+		B: BlockT + Debug,
+	{
+		let recorder = (!self.params.disable_pov_recorder).then(|| Default::default());
+		let backend = sp_state_machine::TrieBackendBuilder::wrap(state.as_trie_backend())
+			.with_optional_recorder(recorder.clone())
+			.build();
+
+		(backend, recorder)
+	}
+}
+
+fn measure_on_block_validation<B: BlockT + Debug>(
+	root: B::Hash,
+	storage_proof: StorageProof,
+	on_validation_batch: Vec<(Vec<u8>, Option<ChildInfo>)>,
+	rounds: u32,
+) -> std::time::Duration {
+	debug!(
+		"POV: len {:?} {:?}",
+		storage_proof.len(),
+		storage_proof.clone().encoded_compact_size::<HashingFor<B>>(root)
+	);
+	let batch_size = on_validation_batch.len();
+	let wasm_module = get_wasm_module();
+	let mut instance = wasm_module.new_instance().unwrap();
+	let params = StorageAccessParams::<B>::new_read(root, storage_proof, on_validation_batch);
+	let dry_run_encoded = params.as_dry_run().encode();
+	let encoded = params.encode();
+
+	let mut durations_in_nanos = Vec::new();
+
+	for i in 1..=rounds {
+		info!("validate_block with {} keys, round {}/{}", batch_size, i, rounds);
+
+		// Dry run to get the time it takes without storage access
+		let dry_run_start = Instant::now();
+		instance.call_export("validate_block", &dry_run_encoded).unwrap();
+		let dry_run_elapsed = dry_run_start.elapsed();
+		debug!("validate_block dry-run time {:?}", dry_run_elapsed);
+
+		let start = Instant::now();
+		instance.call_export("validate_block", &encoded).unwrap();
+		let elapsed = start.elapsed();
+		debug!("validate_block time {:?}", elapsed);
+
+		durations_in_nanos.push((elapsed - dry_run_elapsed).as_nanos() as u64 / batch_size as u64);
```

Comment:
```suggestion
		durations_in_nanos.push(elapsed.saturating_sub(dry_run_elapsed).as_nanos() as u64 / batch_size as u64);
```

Just to be on the safe side.